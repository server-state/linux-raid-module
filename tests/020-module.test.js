const mock = require('mock-fs');
const serverModule = require('../');

describe('Test module export', () => {
    it('should return an error object because no file', async () => {
        // mock file in temporary file system
        mock({
            '/proc/': {
                'mdstat': null
            }
        });
        // apply module function
        const res = await serverModule();
        expect(res).toEqual({error: 'kernel module not loaded'});
    });

    /* it('should return an error object because not parsable file', async () => {
        // mock file in temporary file system
        mock({
            '/proc/': {
                'mdstat': '\n'
            }
        });

        // apply module function
        const res = await serverModule();
        expect(res).toHaveProperty('error', /Please report the issue to module author!/);
    }); */
});

afterAll(() => {
    mock.restore();
});