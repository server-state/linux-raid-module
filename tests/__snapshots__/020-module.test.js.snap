// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test module options Test option 'filter' should filter block device without unique name 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
  ],
}
`;

exports[`Test module options Test option 'filter' should filter everything 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [],
}
`;

exports[`Test module options Test option 'filter' should filter everything except 'myRaid' 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
  ],
}
`;

exports[`Test module options Test option 'filter' should filter everything except 'myRaid' and 'mySecondRaid' 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
  ],
}
`;

exports[`Test module options Test option 'filter' should filter out 'myRaid' 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb9",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb8",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb7",
          "status": "active",
        },
      ],
      "name": "md125",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test module options Test option 'filter' should filter out 'myRaid' and 'mySecondRaid' 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb9",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb8",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb7",
          "status": "active",
        },
      ],
      "name": "md125",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test module options Test option 'filter' should not filter any raids 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb9",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb8",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb7",
          "status": "active",
        },
      ],
      "name": "md125",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test module options Test option 'invert' should do nothing 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb9",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb8",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb7",
          "status": "active",
        },
      ],
      "name": "md125",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test simple module behaviour should return an correct result with one missing unique name 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test simple module behaviour should return an correct result with one wrong link destination 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": undefined,
    },
  ],
}
`;

exports[`Test simple module behaviour should return an result with correct unique names 1`] = `
Object {
  "personalities": Array [
    "raid1",
    "raid0",
  ],
  "raids": Array [
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb3",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb2",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb1",
          "status": "active",
        },
      ],
      "name": "md127",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "myRaid",
    },
    Object {
      "access": "rw",
      "blocks": 62862336,
      "devices": Array [
        Object {
          "index": 2,
          "name": "sdb6",
          "status": "active",
        },
        Object {
          "index": 1,
          "name": "sdb5",
          "status": "active",
        },
        Object {
          "index": 0,
          "name": "sdb4",
          "status": "active",
        },
      ],
      "name": "md126",
      "options": Array [],
      "parameters": "super 1.2 512k chunks",
      "state": "active",
      "type": "raid0",
      "unique": "mySecondRaid",
    },
  ],
}
`;
